### PR TITLE
fix: Improve App Hangs detection

### LINKED ISSUE #1865: Improve App Hangs detection
### Description

With the current App Hangs timeout implementation, it is possible for an App Hangs to last "timeout * 1.9" without being reported.

See https://github.com/getsentry/sentry-cocoa/pull/1861#discussion_r890099824

[Unity implementation](https://github.com/getsentry/sentry-unity/pull/771/files#diff-c0f681085f84da731abf24428667d8fa6285481ebb12bffdf537fa4592ce1c06R122-R142)

### PR DESCRIPTION
**Pull request by kevinrenskers**
## :scroll: Description

See https://github.com/getsentry/sentry-cocoa/issues/1865#issuecomment-1216310438.

## :bulb: Motivation and Context

Closes #1865

## :green_heart: How did you test it?

Existing unit tests, plus manually in the sample app.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

**Review by vaind — COMMENTED** [review]
LGTM

File: Sources/Sentry/SentryANRTracker.m [hunk 1]
```diff
@@ -47,60 +48,33 @@ - (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
...
 
-        if (blockExecutedOnMainThread) {
-            if (wasPreviousANR) {
-                [SentryLog logWithMessage:@"ANR stopped." andLevel:kSentryLevelWarning];
-                [self ANRStopped];
+        if (ticksSinceUiUpdate >= reportTreshold && !reported) {
+            if (![self.crashWrapper isApplicationInForeground]) {
+                [SentryLog logWithMessage:@"Ignoring ANR because the app is in the background"
+                                 andLevel:kSentryLevelDebug];
+                continue;
```
Comment (armcknight) [inline Sources/Sentry/SentryANRTracker.m:71]: Will this spam the logs as long as the app remains backgrounded? If so we could wrap just the log statement in an `if` that checks a boolean like `didLogBgWarning`.

File: Sources/Sentry/SentryANRTracker.m [hunk 2]
```diff
@@ -47,60 +48,33 @@ - (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
 - (void)detectANRs
 {
     NSThread.currentThread.name = @"io.sentry.app-hang-tracker";
-
     self.thread = NSThread.currentThread;
 
-    BOOL wasPreviousANR = NO;
+    __block NSInteger ticksSinceUiUpdate = 0;
+    NSInteger reportTreshold = 5;
+    double sleepIntervalMs = self.timeoutInterval / reportTreshold;
```
[the remark this answers is not in the document]
Comment (kevinrenskers) [reply Sources/Sentry/SentryANRTracker.m:55]: Good point

File: Sources/Sentry/SentryANRTracker.m (hunk 1 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryANRTracker.m:73]: I moved the `self->reported = YES` line up, so now it should only log once.

**Comment by kevinrenskers** [comment]
Hm tests are failing, need to look more into this tomorrow.

**Comment by kevinrenskers** [comment]
Would someone like to look into these tests? Particularly `testMultipleANRs_MultipleReported`, if you add a `print("blockBeforeMainBlock")` inside of the `blockBeforeMainBlock` work, you will see that this gets printed a LOT of times. Like hundreds and hundreds of times. Whereas on the master branch it's like 11 or 12 times. What is going on here?

File: Sources/Sentry/SentryANRTracker.m [hunk 3]
```diff
@@ -47,60 +48,34 @@ - (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
...
-        NSTimeInterval deltaFromNowToBlockDeadline =
-            [[self.currentDate date] timeIntervalSinceDate:blockDeadline];
-
-        if (deltaFromNowToBlockDeadline >= self.timeoutInterval) {
-            NSString *message =
-                [NSString stringWithFormat:@"Ignoring ANR because the delta is too big: %f.",
-                          deltaFromNowToBlockDeadline];
-            [SentryLog logWithMessage:message andLevel:kSentryLevelDebug];
-            continue;
-        }
```
Comment (brustolin) [inline Sources/Sentry/SentryANRTracker.m:93]: h: this check is also missing from the new code

File: Sources/Sentry/SentryANRTracker.m [hunk 4]
```diff
@@ -47,60 +48,34 @@ - (instancetype)initWithTimeoutInterval:(NSTimeInterval)timeoutInterval
...
+        [self.threadWrapper sleepForTimeInterval:sleepInterval];
 
-        [self.threadWrapper sleepForTimeInterval:self.timeoutInterval];
+        if (ticksSinceUiUpdate >= reportTreshold && !reported) {
+            self->reported = YES;
 
-        if (blockExecutedOnMainThread) {
-            if (wasPreviousANR) {
-                [SentryLog logWithMessage:@"ANR stopped." andLevel:kSentryLevelWarning];
-                [self ANRStopped];
```
Comment (brustolin) [inline Sources/Sentry/SentryANRTracker.m:68]: h: you are missing this in the new approach.
End of ANR indication

**Review by brustolin — CHANGES_REQUESTED** [review]
We are missing important part from the old code.

File: Sources/Sentry/SentryANRTracker.m (hunk 3 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryANRTracker.m:93]: The unity code doesn't seem to have something like this either. What kind of check would you expect here? As soon as `ticksSinceUiUpdate` hits 5, we report the ANR, mark it as reported, and that's it. I guess we could check if `ticksSinceUiUpdate` is more than 10, and then ignore it?

File: Sources/Sentry/SentryANRTracker.m (hunk 4 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryANRTracker.m]: Added it back, in a slightly different way. I think it should work 🤞

Comment (brustolin) [reply Sources/Sentry/SentryANRTracker.m]: Not sure about running it in the main thread but I think it works.

**Comment by brustolin** [comment]
Looks good, just need to make CI happy again.

File: Sources/Sentry/SentryANRTracker.m (hunk 3 above)
Comment (kevinrenskers) [reply Sources/Sentry/SentryANRTracker.m]: I had to add that logic back in, to make the `testAppSuspended_NoANR` test pass. And looking at the name of that test, seems important to keep that functionality :)

**Review by brustolin — APPROVED** [review]
LGTM

**Comment by kevinrenskers** [comment]
@armcknight can you re-review? Your previous review is blocking the PR from being merged.

**Review by armcknight — APPROVED** [review]

**Pull request merged**